File: code_review_mcp/src/tools/post-comments.ts
/**
 * MCP Tool: post_pr_comments
 *
 * Posts user-approved review comments as inline comments on a Bitbucket PR.
 * Accepts the PR URL and an array of comments with file path, line number, and content.
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { AppConfig, ReviewComment } from "../config/types.js";
import { parsePrUrl } from "../utils/pr-url-parser.js";
import {
  postInlineComment,
  postGeneralComment,
} from "../providers/bitbucket.js";

export function registerPostPrComments(
  server: McpServer,
  config: AppConfig
): void {
  server.tool(
    "post_pr_comments",
    "Post review comments on a Bitbucket pull request. " +
      "Each comment can be an inline comment (attached to a specific file and line) or a general comment. " +
      "Use this after the user has selected which review comments to post from the get_pr_diff analysis.",
    {
      pr_url: z
        .string()
        .url()
        .describe(
          "The full URL of the Bitbucket pull request, e.g. https://bitbucket.org/workspace/repo/pull-requests/123"
        ),
      comments: z
        .array(
          z.object({
            path: z
              .string()
              .describe(
                "File path relative to repo root. Leave empty string for general (non-inline) comments."
              ),
            line: z
              .number()
              .int()
              .min(0)
              .describe(
                "Line number in the new file version for inline comment. Use 0 for general comments."
              ),
            content: z
              .string()
              .describe(
                "The review comment content in markdown format."
              ),
          })
        )
        .min(1)
        .describe(
          "Array of review comments to post. Each comment has a file path, line number, and content."
        ),
    },
    async ({ pr_url, comments }) => {
      try {
        // 1. Parse the PR URL
        const prInfo = parsePrUrl(pr_url);
        const auth = config.auth.bitbucket;

        // 2. Post each comment
        const results = [];

        for (const comment of comments) {
          const reviewComment: ReviewComment = {
            path: comment.path,
            line: comment.line,
            content: comment.content,
          };

          let result;
          if (!comment.path || comment.line <= 0) {
            // General comment (not attached to a file/line)
            result = await postGeneralComment(
              auth,
              prInfo,
              comment.content
            );
          } else {
            // Inline comment
            result = await postInlineComment(auth, prInfo, reviewComment);
          }

          results.push(result);

          // Small delay between posts to avoid rate limiting
          await sleep(200);
        }

        // 3. Build summary
        const posted = results.filter((r) => r.success).length;
        const failed = results.filter((r) => !r.success).length;

        const summary = {
          total: comments.length,
          posted,
          failed,
          results,
        };

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify(summary, null, 2),
            },
          ],
        };
      } catch (error) {
        const message =
          error instanceof Error ? error.message : String(error);
        return {
          content: [
            {
              type: "text" as const,
              text: `Error posting comments: ${message}`,
            },
          ],
          isError: true,
        };
      }
    }
  );
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: pull_request_maker_mcp/internal/tools/post_comments.go
package tools

import (
	"context"
	"encoding/json"

	"bitbucket.org/mid-kelola-indonesia/pull-request-maker-mcp/internal/bitbucket"
	"bitbucket.org/mid-kelola-indonesia/pull-request-maker-mcp/internal/config"
	"bitbucket.org/mid-kelola-indonesia/pull-request-maker-mcp/internal/urlparser"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// PostPRCommentsInput is the input schema for the post_pr_comments tool.
type PostPRCommentsInput struct {
	PRURL    string            `json:"pr_url" jsonschema:"The full URL of the Bitbucket pull request"`
	Comments []CommentToPost   `json:"comments" jsonschema:"Array of review comments to post"`
}

// CommentToPost represents a single comment to be posted.
type CommentToPost struct {
	Path    string `json:"path" jsonschema:"File path relative to repo root. Empty string for general comments"`
	Line    int    `json:"line" jsonschema:"Line number in the new file version. 0 for general comments"`
	Content string `json:"content" jsonschema:"The review comment content in markdown"`
}

// PostPRCommentsOutput is the structured output returned by post_pr_comments.
type PostPRCommentsOutput struct {
	Total   int                          `json:"total"`
	Posted  int                          `json:"posted"`
	Failed  int                          `json:"failed"`
	Results []bitbucket.PostCommentResult `json:"results"`
}

// RegisterPostPRComments registers the post_pr_comments tool on the MCP server.
func RegisterPostPRComments(server *mcp.Server, cfg *config.AppConfig, client *bitbucket.Client) {
	mcp.AddTool(server,
		&mcp.Tool{
			Name: "post_pr_comments",
			Description: "Post review comments on a Bitbucket pull request. " +
				"Each comment can be an inline comment (attached to a specific file and line) or a general comment. " +
				"Use this after the user has selected which review comments to post from the get_pr_diff or get_branch_diff analysis.",
		},
		func(ctx context.Context, req *mcp.CallToolRequest, input PostPRCommentsInput) (*mcp.CallToolResult, PostPRCommentsOutput, error) {
			prInfo, err := urlparser.ParsePRURL(input.PRURL)
			if err != nil {
				return errorResult(err.Error()), PostPRCommentsOutput{}, nil
			}

			results := make([]bitbucket.PostCommentResult, 0, len(input.Comments))

			for _, c := range input.Comments {
				comment := bitbucket.ReviewComment{
					Path:    c.Path,
					Line:    c.Line,
					Content: c.Content,
				}
				result := client.PostCommentWithDelay(ctx, cfg.Auth.Bitbucket, *prInfo, comment)
				results = append(results, result)
			}

			posted := 0
			failed := 0
			for _, r := range results {
				if r.Success {
					posted++
				} else {
					failed++
				}
			}

			output := PostPRCommentsOutput{
				Total:   len(input.Comments),
				Posted:  posted,
				Failed:  failed,
				Results: results,
			}

			resultJSON, _ := json.Marshal(output)

			return &mcp.CallToolResult{
				Content: []mcp.Content{
					&mcp.TextContent{Text: string(resultJSON)},
				},
			}, output, nil
		},
	)
}
